stash/git.go

```diff
@@ -0,0 +1,486 @@
+	now := time.Now()
+	c.Author.Date = &now
+	if c.Committer != nil {
+		c.Committer.Date = &now
+	}
+
+	obj, err := s.commit(w, r, c)
+	if err != nil {
+		return nil, err
+	}
+
+	var parents []string
+	if obj.ParentHashes == nil {
+		for _, parent := range obj.ParentHashes {
+			parents = append(parents, parent.String())
+		}
+	}
+
+	commit := &Commit{
+		SHA: obj.Hash.String(),
+		Author: &CommitAuthor{
+			Date:  &obj.Author.When,
+			Name:  obj.Author.Name,
+			Email: obj.Author.Email,
+		},
+		Committer: &CommitAuthor{
+			Date:  &obj.Committer.When,
+			Name:  obj.Committer.Name,
+			Email: obj.Committer.Email,
+		},
+		Message:      obj.Message,
+		Tree:         obj.TreeHash.String(),
+		Parents:      parents,
+		PGPSignature: obj.PGPSignature,
+	}
+
+	return commit, nil
+}
+
+// CloneRepository clones the repository at the given URL to the given path.
+// The repository will be cloned into a temporary directory which shall be clean up by the caller.
+func (s *GitService) CloneRepository(ctx context.Context, URL string) (r *git.Repository, dir string, err error) {
+	dir, err = os.MkdirTemp("", "repo-*")
+	if err != nil {
+		return nil, "", err
+	}
+
+	r, err = git.PlainCloneContext(ctx, dir, false, &git.CloneOptions{
+		URL:  URL,
+		Auth: &githttp.BasicAuth{Username: s.Client.username, Password: s.Client.token},
+	})
+	if err != nil {
+		return nil, "", err
+	}
+
+	return r, dir, nil
+}
+
+func (s *GitService) addCommitFiles(w *git.Worktree, dir string, files []CommitFile) error {
+	for _, file := range files {
+		err := writeCommitFile(file, dir)
+		if err != nil {
+			return err
+		}
+		// Adds the new file to the staging area.
+		_, err = w.Add(*file.Path)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func writeCommitFile(file CommitFile, dir string) error {
+	filename := filepath.Join(dir, *file.Path)
+	filePath := strings.Split(*file.Path, "/")
+	if len(filePath) > 1 {
+		fullPath := append([]string{dir}, filePath[0:len(filePath)-1]...)
+		err := os.MkdirAll(strings.Join(fullPath, "/"), 0777)
+		if err != nil {
+			return err
+		}
+	}
+	err := os.WriteFile(filename, []byte(*file.Content), 0644)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Cleanup removes the temporary directory created for the repository.
+func (s *GitService) Cleanup(dir string) error {
+	err := os.RemoveAll(dir)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateBranch creates a new branch with the given name and checkout the branch.
+func (s *GitService) CreateBranch(branchName string, r *git.Repository) error {
+	w, err := r.Worktree()
+	if err != nil {
+		return err
+	}
+
+	branch := fmt.Sprintf("refs/heads/%s", branchName)
+	b := plumbing.ReferenceName(branch)
+
+	// First try to checkout branch
+	err = w.Checkout(&git.CheckoutOptions{Create: false, Force: false, Branch: b})
+
+	if err != nil {
+		// got an error  - try to create it
+		err := w.Checkout(&git.CheckoutOptions{Create: true, Force: false, Branch: b})
```

Comment:
Before doing a create on the branch should you check that the error you've gotten is `plumbing.ErrReferenceNotFound`?

Reply:
`plumbing.ErrReferenceNotFound` is not returned. `checkout` recovers from it when creating.

Reply:
Do you mean `w.Checkout` on line 382 recovers or `w.Checkout` on line 378?

I was referring to line 378, so as well as checking if `err != nil` on line 380 there is another check for `err ==  plumbing.ErrReferenceNotFound` before running `w.Checkout` with `Create: true`.